src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -336,17 +337,38 @@ private HttpMessageHandler SetupHandlerChain()
             CheckDisposed();
             HttpMessageHandler handler = _handler ?? SetupHandlerChain();
 
-            Exception error = ValidateAndNormalizeRequest(request);
+            Exception error = ValidateAndNormalizeRequest(request, out var shouldBufferContent);
             if (error != null)
             {
                 return Task.FromException<HttpResponseMessage>(error);
             }
 
+            if (shouldBufferContent)
+            {
+                return SendWithBufferedContentAsync(handler, request, cancellationToken);
+            }
+
             return handler.SendAsync(request, cancellationToken);
         }
 
-        private Exception ValidateAndNormalizeRequest(HttpRequestMessage request)
+        private async Task<HttpResponseMessage> SendWithBufferedContentAsync(
+            HttpMessageHandler handler, HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            await request.Content.LoadIntoBufferAsync(HttpContent.MaxBufferSize, cancellationToken).ConfigureAwait(false);
+
+            // Forcibly set content length in case caller unset content length.
+            if (request.Content.Headers.ContentLength == null)
```

Comment:
Not sure if we want to actually handle this use-case -- in theory this only happens if the caller purposefully calls `content.Headers.ContentLength = null`, like we currently do in our tests.